### PR TITLE
Let AffineConstraints::make_consistent_in_parallel() iterate until converged.

### PR DESCRIPTION
**Pull request by marcfehling**
Using [`make_consistent_in_parallel`](https://dealii.org/developer/doxygen/deal.II/classAffineConstraints.html#a2d347f2da63fbe1074c571df9f858486) as intended on all locally owned and locally relevant DoFs led to problems in hp-adaptive applications.

Consider for example the following case:
constraints
After calling `make_consistent_in_parallel`, DoFs 16 and 17 on subdomain 0 are constrained against DoF 30, but this DoF is not part of the locally relevant DoFs on subdomain 0.

So if we want to have consistent constraints on all locally relevant DoFs, we need to 
- update the locally stored constraints on the `AffineConstraints` object, and use the IndexSet representing locally stored constraints (`local_lines`) to initialize data structures
- run the exchange of constraints iteratively if locally relevant DoFs are constrained against DoFs we don't know anything about

I've also added the option to perform the exchange of constraints only on locally active DoFs, which should be enough for CG applications. From my understanding of the current algorithm in `internal::compute_locally_relevant_constraints`, this should be possible. Hopefully @peterrum agrees :)

EDIT: The old test `lac/constraints_make_consistent_in_parallel_01.cc` already did exactly that. So this PR only adds the iterative feature.

### CODE REVIEW

**Comment by kronbichler** [comment]
I will look into it soon, for now one question: Does this make #14905 obsolete? I think the present PR should be a more generic solution, making sure the information that arrives is finally merged together. It would still be interesting how this PR works regarding the test case of #14907. When I will look into this PR, I will try to see myself, but I think this PR also addresses those two other cases.

File: include/deal.II/lac/affine_constraints.templates.h [hunk 1]
```diff
@@ -597,36 +597,59 @@ template <typename number>
...
+      // 1) get all locally relevant constraints ("temporal constraint matrix")
+      const auto temporal_constraint_matrix =
+        internal::compute_locally_relevant_constraints(
+          *this,
+          locally_owned_dofs,
+          constraints_to_make_consistent,
+          mpi_communicator);
+
+      // 2) add untracked dofs to the index sets
+      std::vector<types::global_dof_index> constrained_indices;
```
Comment (masterleinad) [inline include/deal.II/lac/affine_constraints.templates.h:621]: Supposedly this can get pretty big for every iteration. Would it make sense to move the variable declaration outside the loop and only `clear()` it here?
Also, it might make sense to `reserve` enough space at least of the size of the `temporal_constraint_matrix` to avoid frequent reallocations in the loop below for the first iteration.

**Comment by marcfehling** [comment]
> Does this make https://github.com/dealii/dealii/pull/14905 obsolete?

I am not sure.

This patch tries to *add* missing constraint lines. It does quietly assume that the constraints are correct on the sending process, and enhances `local_lines` correspondingly so we actually know about missing lines.

If I understood #14905 correctly, that patch is also concerned with constraints being different by supplying a merging strategy. However on first sight, it does not extend `local_lines`.

Both patches assume that chains of constraints can be resolved with `close()` after `make_consistent_in_parallel()`, so they assume that all necessary information is available on each process. Maybe a combination of both patches would be the right way to go?

I can also try out if #14905 helps in my hp-applications.

**Comment by marcfehling** [comment]
@bergbauer -- FYI

File: include/deal.II/lac/affine_constraints.templates.h [hunk 2]
```diff
@@ -597,36 +597,78 @@ template <typename number>
...
+
+  // The IndexSet `constraints_to_make_consistent` contains those DoFs about
+  // which we want to know all constraints. We will receive constraints for
+  // these DoFs against other DoFs, which might be constrained themselves. To
+  // successfully resolve all chains of constraints, we need to know the
+  // constraints of all these DoFs, and we keep track about the relevant DoFs
+  // here.
+  //
+  // This IndexSet stores these DoFs from the previous iteration. If the old and
+  // new index sets match, we have converged.
```
Comment (masterleinad) [inline include/deal.II/lac/affine_constraints.templates.h:626]:
```suggestion
  // This IndexSet keeps track of the locally stored constraints on this
  // AffineConstraints object: If we have received constrained indices that we
  // don't know of, we need to expand our playing field.
  IndexSet locally_stored_constraints;

  // The IndexSet `constraints_to_make_consistent` contains those DoFs about
  // which we want to know all constraints. We will receive constraints for
  // these DoFs against other DoFs, which might be constrained themselves. To
  // successfully resolve all chains of constraints, we need to know the
  // constraints of all these DoFs, and we keep track of the relevant DoFs
  // here.
  //
  // This IndexSet stores these DoFs from the previous iteration. If the old and
  // new index sets match, we have converged.
```

**Comment by bergbauer** [comment]
Thanks for the ping. I am on vacation currently, I will have a look on Monday!

**Review by masterleinad — APPROVED** [review]

**Comment by marcfehling** [comment]
- #14905 does **not** help with the problems in my hp-adaptive applications: Missing constraints had a negative effect on the conditioning of my matrices (large eigenvalues), severely affecting solver performance in a negative way.

- The test from #14907 produces the following output. (To make it run you just need to initialize the Affine Constraints object with the corresponding IndexSets in the current state of this patch.) The output is slightly different to what @peterrum proposed in https://github.com/dealii/dealii/pull/14907/files#r1148609314.
```
DEAL:0::Output 0
DEAL:0::Output 1
    0 1:  0.500000
    0 2:  0.500000
    2 3:  1.50000
DEAL:0::Output 2
    0 1:  0.500000
    0 3:  0.750000
    2 3:  1.50000

DEAL:1::Output 0
    0 1:  0.500000
    0 2:  0.500000
    2 3:  1.50000
DEAL:1::Output 1
    0 1:  0.500000
    0 2:  0.500000
    2 3:  1.50000
DEAL:1::Output 2
    0 1:  0.500000
    0 3:  0.750000
    2 3:  1.50000


DEAL:2::Output 0
    0 1:  0.500000
    0 3:  0.500000
DEAL:2::Output 1
    0 1:  0.500000
    0 2:  0.500000
    2 3:  1.50000
DEAL:2::Output 2
    0 1:  0.500000
    0 3:  0.750000
    2 3:  1.50000
```

**Comment by bangerth** [comment]
I think that exchanging DoF constraints as proposed here is conceptually the right thing to do because it is easy to make mistakes otherwise. That said, in the example you have above, can you explain why it is actually *necessary* to know that 16 and 17 are constrained against 30 (and 26, 28)? At least for CG schemes, we never write into rows 16 and 17 of vectors or matrices, and so never have to the local_to_global operation for these where we would need to know the constraints. For DG schemes, we *would* have to write into these rows and columns (even though these DoFs are on a ghost cell), but in that case 16 and 17 are not actually constrained.

The only case I can imagine is if you use the usual C^0 elements to solve a biharmonic equation using a DG approach to penalize the jump in the gradient. In that case, you get the DG terms from the penalty, and the constraints from the CG discretization.

**Comment by marcfehling** [comment]
> That said, in the example you have above, can you explain why it is actually necessary to know that 16 and 17 are constrained against 30 (and 26, 28)?

The example above is basically just a minimal, previously failing test. It is true that for CG applications, as in the example with `FE_Q` elements, you would only need an exchange of constraints on locally active DoFs. Let me give you a list of events that lead to the proposition of this patch. Maybe that puts things into perspective.

---

In my hp-adaptive applications, I noticed that solver performance was highly dependent on the partitioning of the mesh (determined via checkpointing and restarting with different counts of processes). One idea @kronbichler had was to verify that constraints are consistent across all processes, which in fact was not the case (determined by [`is_consistent_in_parallel`](https://dealii.org/developer/doxygen/deal.II/classAffineConstraints.html#ad5fb8d3ca686a76ba33eb82d7f52b5fd)). So I tried to apply [`make_consistent_in_parallel`](https://dealii.org/developer/doxygen/deal.II/classAffineConstraints.html#a2d347f2da63fbe1074c571df9f858486) with index sets of locally owned and relevant DoFs as suggested by the documentation, but that didn't work:

While resolving chains of constraints [`AffineConstraints::close`](https://dealii.org/developer/doxygen/deal.II/classAffineConstraints.html#a1611aa37f754086388ca76bcd421cce5) complained about lines it doesn't know about, which is what led to the above example and failing test of this PR. For further checks during postprocessing, I also wanted to create sparsity pattern objects with [`DoFTools::make_sparsity_pattern`](https://dealii.org/developer/doxygen/deal.II/group__constraints.html#ga432d35b9ef18d1802d48b576300b04fd) which failed for the same reason. I initialized the sparsity pattern with locally relevant constraints, but then my constraints object suggested that it should know about more DoFs.

So somehow we need to expand the IndexSet of locally relevant DoFs with those DoFs that appear in the new constraint lines. Typically we say that the constraints object's locally stored lines should correspond to the locally relevant DoFs. So what I suggest in this patch is to update the [`local_lines`](https://dealii.org/developer/doxygen/deal.II/classAffineConstraints.html#a7d5ecc66d3618beb896dc162462ae836) member variable. For the initialization of data structures like sparsity patterns or vectors, we would then need to use [`AffineConstraints::get_local_lines`](https://dealii.org/developer/doxygen/deal.II/classAffineConstraints.html#a3a204d1c0a0fdc63c293d102c0adedd1) as the set of locally relevant DoFs.

In addition to resolve the situation that the new constraint lines contain new DoFs that themselves might be constrained, but we do not know their constraint lines yet, we transformed the [`make_consistent_in_parallel`](https://dealii.org/developer/doxygen/deal.II/classAffineConstraints.html#a2d347f2da63fbe1074c571df9f858486) into an iterative process.

---

My hp-adaptive applications now run with the current patch. So far, I have run [`make_consistent_in_parallel`](https://dealii.org/developer/doxygen/deal.II/classAffineConstraints.html#a2d347f2da63fbe1074c571df9f858486) on the locally relevant DoFs, and updated the locally relevant IndexSet by [`AffineConstraints::get_local_lines`](https://dealii.org/developer/doxygen/deal.II/classAffineConstraints.html#a3a204d1c0a0fdc63c293d102c0adedd1) to initialize data structures.

Out of curiosity, I tried to run  [`make_consistent_in_parallel`](https://dealii.org/developer/doxygen/deal.II/classAffineConstraints.html#a2d347f2da63fbe1074c571df9f858486) on the set of locally active DoFs. It works as well, but still requires an update of the locally relevant DoFs by [`AffineConstraints::get_local_lines`](https://dealii.org/developer/doxygen/deal.II/classAffineConstraints.html#a3a204d1c0a0fdc63c293d102c0adedd1) to truly lead to a consistent performance of the solver and the proper initialization of data structures.

**That suggests that in my scenario, there are locally active DoFs with constraints that [`DoFTools::make_hanging_node_constraints`](https://dealii.org/developer/doxygen/deal.II/group__constraints.html#ga3b4ea7dfd313e388d868c4e4aa685799) can not determine on the MPI local process.**

I can spend another day or two to actually find the exact DoF in my scenario that is responsible for all this, but I think the example above and the failing test already depict what is *conceptually* wrong at the moment. I would rather like to move on :-)

**Comment by kronbichler** [comment]
Thank you for the extensive tests and explanations @marcfehling. Based on these arguments, I agree with your assessment that we should accept two things:
- We need an iterative process in `make_consistent_in_parallel` to take into account the information from neighboring processes, say for the case of CG discretization of the biharmonic operator, other constraint types than those from hanging nodes, as those that led @bergbauer to create #14905.
- We need to enlarge the `local_lines` during this process, as constrained degrees of freedom might in fact refer to DoFs not visible in our ghost layer on the present process. Hence, this augmented `local_lines` object needs to be the basis for creating the vectors and matrices.

Seen from a linear algebra perspective, I believe that we are on the right track with the iterative process, because the constraints are, in the end, of algebraic nature that express degrees of freedom by others, and in fact there can be cases where those relations cannot be determined locally from the mesh elements and their single ghost layer available at that process. I guess simple h-refinement has lured us for too long that a local operation is sufficient.

**Comment by kronbichler** [comment]
> So right now it appears to me that we need both #14905 and #17175.

I agree, based on what you write, that we should see over that other patch again. Thank you for the test and checking @marcfehling. I think it would be good to form some kind of decision on that old patch #14905 for the present release, before we keep forgetting about it and accumulate more items. I would appreciate for @bergbauer to have another look when he's back from holiday and have a discussion on that topic, and I realize that these two approaches seem to address different topics.

File: include/deal.II/lac/affine_constraints.h [hunk 3]
```diff
@@ -1990,13 +1990,37 @@ class AffineConstraints : public Subscriptor
...
-   * in a distributed computation. One should call this function before
-   * calling close().
+   * Make the constraints that are locally stored on the current object
+   * consistent on all MPI processes in a distributed computation.
+   *
+   * The IndexSet @p locally_owned_dofs conforms to all DoFs that are locally
+   * owned on the current process, i.e., DoFHandler::locally_owned_dofs().
+   *
+   * The IndexSet @p constraints_to_make_consistent shall contain all DoF
+   * indices on which we want to store the complete constraints for. Which DoFs
```
Comment (kronbichler) [inline include/deal.II/lac/affine_constraints.h:2000]: I am not sure in terms of the preposition, but I think it sounds better with
```suggestion
   * indices for which we want to store the complete constraints for. Which DoFs
```

File: include/deal.II/lac/affine_constraints.h [hunk 4]
```diff
@@ -1990,13 +1990,37 @@ class AffineConstraints : public Subscriptor
...
+   * these might be depends on your actual use case: For most applications in
+   * which you are using continuous Galerkin methods, you would want to use
+   * locally active DoFs here, obtained via
+   * DoFTools::extract_locally_active_dofs(). However in discontinuous Galerkin
+   * methods, you might need consistent information on all locally relevant
+   * DoFs, for which you need DoFTools::extract_locally_relevant_dofs().
```
Comment (kronbichler) [inline include/deal.II/lac/affine_constraints.h:2006]:
```suggestion
   * DoFs due to the need to compute face integrals against ghosted cells,
   * for which you need DoFTools::extract_locally_relevant_dofs().
```

File: include/deal.II/lac/affine_constraints.h [hunk 5]
```diff
@@ -1990,13 +1990,37 @@ class AffineConstraints : public Subscriptor
...
+   *
+   * This function updates the locally stored constraints (or local_lines) of
+   * this object, whenever a DoF is constrained against another that we have not
+   * known of before. After calling this function, it might be necessary to use
+   * the IndexSet retrieved by get_local_lines() when initializing data
+   * structures, for example:
```
Comment (kronbichler) [inline include/deal.II/lac/affine_constraints.h:2012]:
```suggestion
   * the IndexSet retrieved by get_local_lines() when initializing data
   * structures built on top of the AffineConstraints object, for example:
```

**Review by kronbichler — APPROVED** [review]

File: include/deal.II/lac/affine_constraints.h [hunk 6]
```diff
@@ -1990,13 +1990,38 @@ class AffineConstraints : public Subscriptor
...
+   * LinearAlgebra::distributed::Vector<double> solution(
+   *   locally_owned_dofs, constraints.get_local_lines(), mpi_communicator);
+   * @endcode
+   *
+   * @note One should call this function before calling close().
    */
   void
   make_consistent_in_parallel(const IndexSet &locally_owned_dofs,
-                              const IndexSet &locally_relevant_dofs,
+                              IndexSet        constraints_to_make_consistent,
```
Comment (peterrum) [inline include/deal.II/lac/affine_constraints.h:2024]: Could we revert this change?

File: include/deal.II/lac/affine_constraints.templates.h [hunk 7]
```diff
@@ -597,36 +597,79 @@ template <typename number>
...
+  // these DoFs against other DoFs, which might be constrained themselves. To
+  // successfully resolve all chains of constraints, we need to know the
+  // constraints of all these DoFs, and we keep track of the relevant DoFs
+  // here.
+  //
+  // This IndexSet stores these DoFs from the previous iteration. If the old and
+  // new index sets match, we have converged.
+  IndexSet constraints_made_consistent;
+
+  do
```
Comment (peterrum) [inline include/deal.II/lac/affine_constraints.templates.h:630]: I think we should introduce a counter somewhere to prevent deadlock (?).

**Comment by marcfehling** [comment]
This patch should solve the issue I described in https://github.com/dealii/dealii/issues/6255#issuecomment-1034233517 (but it won't solve the core issue of #6255). I will run some more tests today.

---

I created a [branch](https://github.com/marcfehling/dealii/tree/constraints-combined) that combines #14905 and this patch. The test from #14907 now triggers an assertion. I am afraid that both patches can't co-exist peacefully.
```
--------------------------------------------------------
An error occurred in line <107> of file <.../dealii/include/deal.II/lac/affine_constraints.templates.h> in function
    void dealii::AffineConstraints<number>::add_constraint(size_type, const dealii::ArrayView<const std::pair<unsigned int, number> >&, number) [with number = double; size_type = unsigned int]
The violated condition was: 
    std::adjacent_find(column_indices.begin(), column_indices.end()) == column_indices.end()
Additional information: 
    You are trying to insert a constraint that lists the same degree of
    freedom more than once on the right hand side. This is not allowed.
--------------------------------------------------------
```

**Comment by kronbichler** [comment]
> I am afraid that both patches [this one and #14905] can't co-exist peacefully.

I don't think that's necessarily a bad result, it might be that we need to pick #14905 up again and reconsider what @bergbauer and I did there, I would be happy to give it another thought once this work is complete (which I highly appreciate, I think it makes things easier).

**Comment by peterrum** [comment]
This should fix the issue:
```diff
diff --git a/include/deal.II/lac/affine_constraints.templates.h b/include/deal.II/lac/affine_constraints.templates.h
index dc1799f5f1..1bf8d9d9fd 100644
--- a/include/deal.II/lac/affine_constraints.templates.h
+++ b/include/deal.II/lac/affine_constraints.templates.h
@@ -327,6 +327,13 @@ namespace internal
       if (locally_relevant_constraints.empty())
         return;
 
+      for(auto & entry: locally_relevant_constraints)
+      std::sort(
+        entry.entries.begin(),
+        entry.entries.end(),
+        [](const auto &l1,
+           const auto &l2) { return l1.first < l2.first; });
+
       std::sort(
         locally_relevant_constraints.begin(),
         locally_relevant_constraints.end(),
@@ -387,7 +394,31 @@ namespace internal
               // merge entries vectors if different, otherwise ignore the
               // second entry
               if (!vectors_are_equal)
-                av.insert(av.end(), bv.begin(), bv.end());
+                {
+                  const auto sizea = av.size();
+                  const auto sizeb = bv.size();
+
+                  // TODO: use "merge" of "merge sort" to prevent quadratic
+                  // complexity
+                  for (unsigned int j = 0; j < sizeb; ++j)
+                  {
+                    bool new_entry = true;
+
+                    for (unsigned int i = 0; i < sizea; ++i)
+                      {
+                        if (av[i].first == bv[j].first)
+                        {
+                          Assert(equal_with_tol(av[i].second, bv[j].second),
+                             ExcInternalError());
+
+                          new_entry = false;
+                        }
+                      }
+
+                    if(new_entry)
+                      av.push_back(bv[j]);
+                  }
+                }
             }
         }
       ++write_ptr;
diff --git a/tests/mpi/constraints_consistent_02.mpirun=3.output b/tests/mpi/constraints_consistent_02.mpirun=3.output
index c586925ab6..b763477226 100644
--- a/tests/mpi/constraints_consistent_02.mpirun=3.output
+++ b/tests/mpi/constraints_consistent_02.mpirun=3.output
@@ -2,8 +2,8 @@
 DEAL:0::Output 0
 DEAL:0::Output 1
     0 1:  0.500000
-    0 3:  0.500000
     0 2:  0.500000
+    0 3:  0.500000
     2 3:  1.50000
 DEAL:0::Output 2
     0 1:  0.500000
@@ -16,8 +16,8 @@ DEAL:1::Output 0
     2 3:  1.50000
 DEAL:1::Output 1
     0 1:  0.500000
-    0 3:  0.500000
     0 2:  0.500000
+    0 3:  0.500000
     2 3:  1.50000
 DEAL:1::Output 2
     0 1:  0.500000
@@ -30,8 +30,8 @@ DEAL:2::Output 0
     0 3:  0.500000
 DEAL:2::Output 1
     0 1:  0.500000
-    0 3:  0.500000
     0 2:  0.500000
+    0 3:  0.500000
     2 3:  1.50000
 DEAL:2::Output 2
     0 1:  0.500000
```

When merging the constraints, we did not removed duplicates.

**Comment by bergbauer** [comment]
@peterrum was faster than me. The problem is with the combined patches that in the second call to `internal::compute_locally_relevant_constraints` we get a duplicated entry due to the enlarged `IndexSet`.

> and I realize that these two approaches seem to address different topics.

I agree with @kronbichler s assessment that the two patches address different topics. Thanks for tackling this @marcfehling , this will also help in my application case with chains of constraints over several partitions of the mesh.

**Comment by marcfehling** [comment]
Thank you all for your feedback! I will squash this PR to get it in shape for merging.

Of course there are ways on how to do things more efficient in this patch. I opted to use many separate IndexSets to keep track of changed DoFs and used the `AffineConstraints::reinit` function. One could also work directly with `this->local_lines` at some places, but in favor of readability I left it like this.

However, you will notice that I had to adjust test [constraints_make_consistent_in_parallel_01.cc](https://github.com/dealii/dealii/pull/17175/files#diff-7565b469fce5079eddfccfc51d4dfa3ba5b5273db3ebc8ea7508dfd0aa97979c). This whole approach only works when `AffineConstraints` is actually initialized with `locally_stored_constraints`. So my question would be how to handle this case:
- Add an assertion to check whether `this->local_lines` is not empty, i.e., force the user to initialize `AffineConstraints` correctly.
- Skip updating `this->local_lines` if it is empty, i.e., consider the extension of `this->local_lines` as a feature.

What's your opinion?

**Review by peterrum — APPROVED** [review]

**Comment by marcfehling** [comment]
> Can we add an assert?

I have added an assertion.

**Review by peterrum — APPROVED** [review]
